### PR TITLE
core: add support for dogstatsd unix socket

### PR DESCRIPTION
**Pull request by majorgreys**
Support configuring dogstatsd client in tracer with a unix socket

### CODE REVIEW

File: ddtrace/tracer.py [hunk 1]
```diff
@@ -142,8 +155,8 @@ def context_provider(self):
 
     # TODO: deprecate this method and make sure users create a new tracer if they need different parameters
     def configure(self, enabled=None, hostname=None, port=None, uds_path=None, https=None,
-                  dogstatsd_host=None, dogstatsd_port=None, sampler=None, context_provider=None,
```
Comment (brettlangdon) [inline ddtrace/tracer.py]: we can't remove `dogstatsd_host` and `dogstatsd_port` from here, breaks the API.

Comment (brettlangdon) [reply ddtrace/tracer.py]: it is also tricky, because if they did supply `hostname='value'` I wonder if we should inherit that for dogstatsd as well.

Comment (majorgreys) [reply ddtrace/tracer.py]: 👍

File: ddtrace/tracer.py [hunk 2]
```diff
@@ -22,6 +22,20 @@
 log = get_logger(__name__)
 
 
+def _parse_dogstatsd_url(dogstatsd_url):
+    if dogstatsd_url is not None:
```
Comment (brettlangdon) [inline ddtrace/tracer.py]: can clean this up:

```python
if dogstatsd_url is None:
    return

....
```

Make it easier than reading an entire indented portion

File: ddtrace/tracer.py [hunk 3]
```diff
@@ -187,16 +201,13 @@ def configure(self, enabled=None, hostname=None, port=None, uds_path=None, https
...
-        self.log.debug('Connecting to DogStatsd on {}:{}'.format(
-            self._dogstatsd_host,
-            self._dogstatsd_port,
-        ))
-        self._dogstatsd_client = DogStatsd(
-            host=self._dogstatsd_host,
-            port=self._dogstatsd_port,
-        )
+        if dogstatsd_host is not None and dogstatsd_port is not None and dogstatsd_url is None:
+            dogstatsd_url = '{}:{}'.format(dogstatsd_host, dogstatsd_port)
```
Comment (brettlangdon) [inline ddtrace/tracer.py]:
```python
log.warning('deprecated: please use dogstatsd_url="%s:%d" instead' % (dogstatsd_host, dogstatsd_port))
```

Comment (brettlangdon) [reply ddtrace/tracer.py]: (or a real deprecation warning is fine too, not just a log)

Comment (majorgreys) [reply ddtrace/tracer.py]: 👍

File: ddtrace/tracer.py [hunk 4]
```diff
@@ -187,16 +203,15 @@ def configure(self, enabled=None, hostname=None, port=None, uds_path=None, https
...
-            self._dogstatsd_port,
-        ))
-        self._dogstatsd_client = DogStatsd(
-            host=self._dogstatsd_host,
-            port=self._dogstatsd_port,
-        )
+        if dogstatsd_host is not None and dogstatsd_url is None:
+            dogstatsd_url = '{}:{}'.format(dogstatsd_host, dogstatsd_port or self.DEFAULT_DOGSTATSD_PORT)
+            warn(('tracer.configure(): dogstatsd_host and dogstatsd_port are deprecated. '
+                  'Use dogstatsd_url').format(dogstatsd_url))
```
Comment (brettlangdon) [inline ddtrace/tracer.py]:
```suggestion
                  'Use dogstatsd_url={!r}').format(dogstatsd_url))
```

**Comment by skaji** [comment]
I don't think we should call `host:port` as *URL*.
If we use "DD_DOGSTATSD_URL" environment variable, how about using `udp://host:port` form ?

File: ddtrace/tracer.py [hunk 5]
```diff
@@ -14,14 +14,30 @@
...
+    if dogstatsd_url is None:
+        return
+
+    # url can be either of the form `<host>:<port>` or `unix://<path>`
+    if dogstatsd_url.startswith('unix://'):
+        path = dogstatsd_url[len('unix://'):]
+        return dict(socket_path=path)
+    else:
+        try:
+            hostname, port = dogstatsd_url.split(':')
```
Comment (jd) [inline ddtrace/tracer.py]: You say it's an URL and then just treat that as a `hostname:port`?

Anyhow, this is unlikely to work with IPv6.

File: ddtrace/tracer.py [hunk 6]
```diff
@@ -14,14 +14,30 @@
...
+            return dict(host=hostname, port=int(port))
+        except ValueError:
+            raise ValueError('Unknown url format for `{}`'.format(dogstatsd_url))
```
Comment (jd) [inline ddtrace/tracer.py]: URL

File: ddtrace/tracer.py [hunk 7]
```diff
@@ -14,14 +14,30 @@
...
 log = get_logger(__name__)
 
 
+def _parse_dogstatsd_url(dogstatsd_url):
+    if dogstatsd_url is None:
+        return
+
+    # url can be either of the form `<host>:<port>` or `unix://<path>`
+    if dogstatsd_url.startswith('unix://'):
+        path = dogstatsd_url[len('unix://'):]
```
Comment (jd) [inline ddtrace/tracer.py]: I'd suggest something like:

```python
# Support a path directly
if dogstatsd_url.startswith('/'):
    dogstatsd_url = 'unix://' + dogstatsd_url
elif '://' not in dogstats_url:
    dogstatsd_url = 'udp://' + dogstatsd_url

parsed = compat.parse.urlparse(dogstatsd_url)
# check parsed.scheme and act accordingly

```

File: ddtrace/tracer.py [hunk 8]
```diff
@@ -187,16 +203,15 @@ def configure(self, enabled=None, hostname=None, port=None, uds_path=None, https
...
-        self._dogstatsd_port = dogstatsd_port or self._dogstatsd_port
-        self.log.debug('Connecting to DogStatsd on {}:{}'.format(
-            self._dogstatsd_host,
-            self._dogstatsd_port,
-        ))
-        self._dogstatsd_client = DogStatsd(
-            host=self._dogstatsd_host,
-            port=self._dogstatsd_port,
-        )
+        if dogstatsd_host is not None and dogstatsd_url is None:
```
Comment (jd) [inline ddtrace/tracer.py]: Then you realize it's not possible to then only pass a port without `dogstatsd_host` being set.

Gosh I hate this method 😠 😆

Comment (majorgreys) [reply ddtrace/tracer.py]: I added a deprecation warning for those parameters, which I agree we should remove.

File: ddtrace/tracer.py [hunk 9]
```diff
@@ -38,9 +59,10 @@ class Tracer(object):
     DEFAULT_HOSTNAME = environ.get('DD_AGENT_HOST', environ.get('DATADOG_TRACE_AGENT_HOSTNAME', 'localhost'))
     DEFAULT_PORT = int(environ.get('DD_TRACE_AGENT_PORT', 8126))
     DEFAULT_DOGSTATSD_PORT = int(get_env('dogstatsd', 'port', 8125))
+    DEFAULT_DOGSTATSD_URL = get_env('dogstatsd', 'url', '{}:{}'.format(DEFAULT_HOSTNAME, DEFAULT_DOGSTATSD_PORT))
```
Comment (brettlangdon) [inline ddtrace/tracer.py]:
```suggestion
    DEFAULT_DOGSTATSD_URL = get_env('dogstatsd', 'url', 'udp://{}:{}'.format(DEFAULT_HOSTNAME, DEFAULT_DOGSTATSD_PORT))
```

?

File: ddtrace/tracer.py [hunk 10]
```diff
@@ -187,16 +208,15 @@ def configure(self, enabled=None, hostname=None, port=None, uds_path=None, https
...
-        self.log.debug('Connecting to DogStatsd on {}:{}'.format(
-            self._dogstatsd_host,
-            self._dogstatsd_port,
-        ))
-        self._dogstatsd_client = DogStatsd(
-            host=self._dogstatsd_host,
-            port=self._dogstatsd_port,
-        )
+        if dogstatsd_host is not None and dogstatsd_url is None:
+            dogstatsd_url = '{}:{}'.format(dogstatsd_host, dogstatsd_port or self.DEFAULT_DOGSTATSD_PORT)
```
Comment (brettlangdon) [inline ddtrace/tracer.py]:
```suggestion
            dogstatsd_url = 'udp://{}:{}'.format(dogstatsd_host, dogstatsd_port or self.DEFAULT_DOGSTATSD_PORT)
```

?

File: ddtrace/tracer.py [hunk 11]
```diff
@@ -165,6 +186,9 @@ def configure(self, enabled=None, hostname=None, port=None, uds_path=None, https
         :param priority_sampling: enable priority sampling, this is required for
             complete distributed tracing support. Enabled by default.
         :param collect_metrics: Whether to enable runtime metrics collection.
+        :param str dogstatsd_host: Host for UDP connection to DogStatsD (deprecated: use dogstatsd_url)
```
Comment (majorgreys) [inline ddtrace/tracer.py]: @brettlangdon added deprecated warning to docstring as well

**Pull request merged**